### PR TITLE
feat: add `sequelize.withConnection`, rename `sequelize.set` to `sequelize.setSessionVariables`

### PR DESCRIPTION
**Pull request by ephys**
## Pull Request Checklist

- [x] Have you added new tests to prevent regressions?
- [x] If a documentation update is necessary, have you opened a PR to [the documentation repository](https://github.com/sequelize/website/)? 
- [x] Did you update the typescript typings accordingly (if applicable)?
- [x] Does the description below contain a link to an existing issue (Closes #[issue]) or a description of the issue you are solving?
- [x] Does the name of your PR follow [our conventions](https://github.com/sequelize/sequelize/blob/main/CONTRIBUTING.md#6-commit-your-modifications)?

## Description Of Change

Closes https://github.com/sequelize/sequelize/issues/15388

This PR implements `sequelize.withConnection`. It's a method you can call to reserve a connection and use it multiple times before releasing it back to the pool.

This is useful in dialects that support session variables & session options.

It is now possible to specify on which connection a query must run by using the `connection` option. This option is mutually exclusive with the `transaction` option.

```ts
sequelize.withConnection(async (connection) => {
  await sequelize.query('SET SESSION group_concat_max_len = 10485760;', { connection });
  await User.findAll({ connection });
});
```

This `connection` option takes precedence over CLS transactions. Unlike transactions, the connection reserved by `withConnection` is not (currently) passed around using CLS.

---

This PR also renames `sequelize.set` to `sequelize.setSessionVariables`, to clarify what it does. It also adds support for CLS transactions in `setSessionVariables`.

---

This PR is a blocker for https://github.com/sequelize/sequelize/pull/15842

### CODE REVIEW

File: packages/core/src/sequelize.js [hunk 1]
```diff
@@ -743,25 +743,23 @@ Use Sequelize#query if you wish to use replacements.`);
...
    */
-  async set(variables, options) {
-
+  async setSessionVariables(variables, options) {
     // Prepare options
-    options = { ...this.options.set, ...typeof options === 'object' && options };
+    options = { ...this.options.set, ...options };
 
     if (!['mysql', 'mariadb'].includes(this.options.dialect)) {
       throw new Error('sequelize.set is only supported for mysql or mariadb');
```
Comment (WikiRik) [inline packages/core/src/sequelize.js:756]:
```suggestion
      throw new Error('sequelize.setSessionVariables is only supported for mysql or mariadb');
```

File: packages/core/src/sequelize.js [hunk 2]
```diff
@@ -743,25 +743,23 @@ Use Sequelize#query if you wish to use replacements.`);
...
     }
 
-    if (!options.transaction || !(options.transaction instanceof Transaction)) {
-      throw new TypeError('options.transaction is required');
+    setTransactionFromCls(options, this);
+
+    if ((!options.transaction || !(options.transaction instanceof Transaction)) && (!options.connection)) {
+      throw new Error('You must specify either options.transaction or options.connection, as sequelize.set is used to set the session options of a connection');
```
Comment (WikiRik) [inline packages/core/src/sequelize.js:762]:
```suggestion
      throw new Error('You must specify either options.transaction or options.connection, as sequelize.setSessionVariables is used to set the session options of a connection');
```

File: packages/core/src/sequelize.js [hunk 3]
```diff
@@ -743,25 +743,23 @@ Use Sequelize#query if you wish to use replacements.`);
...
-   * @param {object}        [options] query options.
-   * @param {Transaction}   [options.transaction] The transaction that the query should be executed under
-   *
-   * @memberof Sequelize
+   * @param {object} variables Object with multiple variables.
+   * @param {object} [options] query options.
    *
    * @returns {Promise}
    */
-  async set(variables, options) {
```
Comment (WikiRik) [inline packages/core/src/sequelize.js:754]: Don't we want to deprecate `sequelize.set` first (aka `set` is an alias for `setSessionVariables`) and remove it later? Or is it used too little to need it?

File: packages/core/test/integration/sequelize/set-session-variables.test.ts [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+      const [data] = await clsSequelize.query<{ foo: string }>('SELECT @foo as `foo`', { type: QueryTypes.SELECT });
+      expect(data).to.be.ok;
+      expect(data.foo).to.be.equal('bar');
+    });
+  });
+
+  it('supports manual transactions', async () => {
+    const transaction = await sequelize.startUnmanagedTransaction();
+
+    try {
```
Comment (WikiRik) [inline packages/core/test/integration/sequelize/set-session-variables.test.ts:41]: Don't think we need this try-catch here, do we?

File: packages/core/src/sequelize.js [hunk 5]
```diff
@@ -743,25 +743,23 @@ Use Sequelize#query if you wish to use replacements.`);
...
-
+  async setSessionVariables(variables, options) {
     // Prepare options
-    options = { ...this.options.set, ...typeof options === 'object' && options };
+    options = { ...this.options.set, ...options };
```
Comment (WikiRik) [inline packages/core/src/sequelize.js:753]: Do we want to keep using `options.set` or do we also want to rename it to something more descriptive?

Comment (ephys) [reply packages/core/src/sequelize.js:753]: It should be aligned with the name of the method so setSessionVariables

File: packages/core/test/integration/sequelize/set-session-variables.test.ts (hunk 4 above)
Comment (ephys) [reply packages/core/test/integration/sequelize/set-session-variables.test.ts:41]: It should never happen but if this throws before the call to commit, we'll have an uncommitted transaction, and that can have an impact on subsequent tests

File: packages/core/src/sequelize.js (hunk 3 above)
Comment (ephys) [reply packages/core/src/sequelize.js:754]: Considering we're also renaming the global option, I think it's safer to not deprecate it and document the breaking change instead

File: packages/core/src/sequelize.js (hunk 1 above)
Comment (ephys) [reply packages/core/src/sequelize.js]: Done

File: packages/core/src/sequelize.js (hunk 2 above)
Comment (ephys) [reply packages/core/src/sequelize.js]: Done

**Review by WikiRik — APPROVED** [review]

**Pull request merged**